pkg/distributor/distributor.go

```diff
@@ -814,16 +817,20 @@ func (d *Distributor) prePushRelabelMiddleware(next push.Func) push.Func {
 		}
 
 		var removeTsIndexes []int
+		lb := labels.NewBuilder(labels.EmptyLabels())
 		for tsIdx := 0; tsIdx < len(req.Timeseries); tsIdx++ {
 			ts := req.Timeseries[tsIdx]
 
 			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
-				l, keep := relabel.Process(mimirpb.FromLabelAdaptersToLabels(ts.Labels), mrc...)
+				lb.Reset(mimirpb.FromLabelAdaptersToLabels(ts.Labels))
+				lb.Set(metaLabelTenantID, userID)
+				keep := relabel.ProcessBuilder(lb, mrc...)
 				if !keep {
 					removeTsIndexes = append(removeTsIndexes, tsIdx)
 					continue
 				}
-				ts.Labels = mimirpb.FromLabelsToLabelAdapters(l)
+				lb.Del(metaLabelTenantID)
+				ts.Labels = mimirpb.FromLabelsToLabelAdapters(lb.Labels(labels.EmptyLabels()))
```

Comment:
This change should make it slightly more performant. Since we want to overwrite `ts.Labels` anyway then we can pass it:

```suggestion
				ts.Labels = mimirpb.FromLabelsToLabelAdapters(lb.Labels(ts.Labels))
```

Reply:
But `ts.Labels` is `[]LabelAdapter` and `lb.Labels()` wants Labels

Reply:
Don't worry about this. New `stringlabels` version of labels code that we will use soon doesn't even take this parameter anymore.

Reply:
> Don't worry about this too much. New `stringlabels` version of labels code that we will use soon doesn't even take this parameter anymore.

Sorry, it's not "new `stringlabels` version" that doesn't have this method with `labels.Labels` parameter anymore, but also latest Prometheus main (after merging https://github.com/prometheus/prometheus/pull/12173) which will be in Mimir after https://github.com/grafana/mimir/pull/4759 gets merged.